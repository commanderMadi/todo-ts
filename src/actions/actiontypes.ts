import {
  GetCategoriesAction,
  DeleteTaskAction,
  AddTaskAction,
  AuthenticationAction
} from '../actions';

export enum ActionTypes {
  getCategories,
  deleteTask,
  addTask,
  checkAuthentication
}

export type Action =
  | GetCategoriesAction
  | DeleteTaskAction
  | AddTaskAction
  | AuthenticationAction;
