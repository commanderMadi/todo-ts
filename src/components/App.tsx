import React from 'react';
import { Link } from 'react-router-dom';
class App extends React.Component {
    render() {
        return (
            <div>
                <h1>Welcome to the app!</h1>
                <Link to='/tasks'>Go to the Tasks page</Link>
            </div>
        );
    }
}

export { App };
