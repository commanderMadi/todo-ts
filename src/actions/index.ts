import axios from 'axios';
import { Dispatch } from 'redux';
import { ActionTypes } from './actiontypes';

const url = 'data.json';

export interface Category {
    name: string;
    todos: Todo[];
}

export interface Todo {
    id: string;
    title: string;
    categoryTitle?: string;
    error?: null;
}

export interface GetCategoriesAction {
    type: ActionTypes.getCategories;
    payload: Category[];
}

export interface DeleteTodoAction {
    type: ActionTypes.deleteTodo;
    payload: { id: string };
}

export interface AddTodoAction {
    type: ActionTypes.addTodo;
    payload: Todo;
}

export const getCategories = () => {
    return async (dispatch: Dispatch) => {
        const response = await axios.get<Category[]>(url);

        dispatch<GetCategoriesAction>({
            type: ActionTypes.getCategories,
            payload: response.data
        });
    };
};

export const deleteTodo = (id: string) => {
    return {
        type: ActionTypes.deleteTodo,
        payload: { id }
    };
};

export const addTodo = (newTodo: Todo): AddTodoAction => {
    return {
        type: ActionTypes.addTodo,
        payload: newTodo
    };
};
