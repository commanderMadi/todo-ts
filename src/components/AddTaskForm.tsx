import React from 'react';
import { Task, Category } from '../actions';
import uuid from 'uuid';
import { FormContainer, Form } from './styles/forms';
import { Button } from '../components/styles/base';

interface AddTaskFormProps {
  categories: Category[];
  onSubmit(newTask: Task): void;
}

export class AddTaskForm extends React.Component<AddTaskFormProps> {
  constructor(props: AddTaskFormProps) {
    super(props);

    this.onSubmit = this.onSubmit.bind(this);
  }
  state: Task = {
    id: 'default-id',
    title: '',
    categoryTitle: '',
    error: null
  };

  onSubmit(e: any) {
    e.preventDefault();
    let title = e.target.elements[0].value;
    let categoryTitle = e.target.elements[1].value;
    let id = uuid();
    if (title && categoryTitle !== 'Select an option') {
      this.setState(
        () => {
          return {
            title,
            categoryTitle,
            id,
            error: null
          };
        },
        () => this.props.onSubmit(this.state)
      );
      e.target.elements[0].value = '';
      e.target.elements[1].value = 'Select an option';
    } else if (
      !title ||
      !categoryTitle ||
      categoryTitle === 'Select an option'
    ) {
      this.setState(() => {
        return {
          error:
            'Please make sure you you have entered a Task title and selected a category'
        };
      });
    }
  }
  render() {
    return (
      <FormContainer>
        {this.state.error && (
          <p className='col-md-12 text-danger mt-4'>{this.state.error}</p>
        )}
        <h4 className='mt-2'>Add a New Task</h4>
        <Form className='mt-4' onSubmit={this.onSubmit}>
          <input
            className='form-control'
            id='task_title'
            placeholder='Enter task title...'
            type='text'
          />
          <select id='formcontrolselect' className='form-control'>
            <option>Select an option</option>
            {this.props.categories.map((category, i: number) => {
              return <option key={i}>{category.name}</option>;
            })}
          </select>
          <Button className='btn btn-primary' type='submit'>
            Add Task
          </Button>
        </Form>
      </FormContainer>
    );
  }
}
