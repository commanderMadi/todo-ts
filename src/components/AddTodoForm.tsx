import React from 'react';
import { connect } from 'react-redux';
import { Todo, addTodo, Category } from '../actions';

interface AddTodoFormProps {
    categories: Category[];
    addTodo: typeof addTodo;
}

class AddTodoForm extends React.Component<AddTodoFormProps> {
    state: Todo = {
        id: 0,
        title: '',
        categoryTitle: ''
    };

    handleInputChange = (e: any) => {
        const todoTitle = e.target.value;

        let id = Math.floor(Math.random() * 100);
        this.setState(() => {
            return {
                title: todoTitle,
                id
            };
        });
    };
    handleSelectChange = (e: any) => {
        e.preventDefault();
        const todoCategoryTitle = e.target.value;
        this.setState(() => {
            return {
                categoryTitle: todoCategoryTitle
            };
        });
    };

    onAddTodo = (e: any) => {
        console.log(this.state);
        e.preventDefault();
        this.props.addTodo(this.state);
    };
    render() {
        return (
            <form onSubmit={this.onAddTodo}>
                <label>Todo Title:</label>
                <input type='text' onChange={this.handleInputChange}></input>
                <label>Todo Category:</label>
                <select onChange={this.handleSelectChange}>
                    <option value='null'>Select an option</option>
                    {this.props.categories.map((category, i: number) => {
                        return <option key={i}>{category.name}</option>;
                    })}
                </select>
                <input type='submit' value='Add Todo' />
            </form>
        );
    }
}

export default connect(undefined, { addTodo })(AddTodoForm);
