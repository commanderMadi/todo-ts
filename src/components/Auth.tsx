import React from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { Auth } from '../actions';
import { ReduxStoreState } from '../reducers';

interface AuthenticateProps {
  isAuthenticated: Auth;
}

export const requireAuth = (CurrentComponent: React.ComponentType<any>) => {
  class Authenticate extends React.Component<AuthenticateProps> {
    render() {
      const { isAuthenticated } = this.props;
      const { auth } = isAuthenticated;
      return !auth ? (
        <Redirect to='/login' />
      ) : (
        <CurrentComponent {...this.props} />
      );
    }
  }

  const mapStateToProps = (state: ReduxStoreState) => ({
    isAuthenticated: state.isAuthenticated
  });
  return connect(mapStateToProps)(Authenticate);
};

export const publicRoute = (CurrentComponent: React.ComponentType<any>) => {
  class PublicRoute extends React.Component<AuthenticateProps> {
    render() {
      const { isAuthenticated } = this.props;
      const { auth } = isAuthenticated;
      return !auth ? (
        <CurrentComponent {...this.props} />
      ) : (
        <Redirect to='/profile' />
      );
    }
  }

  const mapStateToProps = (state: ReduxStoreState) => ({
    isAuthenticated: state.isAuthenticated
  });
  return connect(mapStateToProps)(PublicRoute);
};
