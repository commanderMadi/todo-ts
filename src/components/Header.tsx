import React from 'react';
import { Link, NavLink } from 'react-router-dom';
import { ReduxStoreState } from '../reducers';
import { Auth, checkAuthentication } from '../actions';
import { connect } from 'react-redux';

export interface HeaderProps {
    isAuthenticated: Auth;
    checkAuthentication: typeof checkAuthentication;
}

export class Header extends React.Component<HeaderProps> {
    onLogoutclick = () => {
        this.props.checkAuthentication(this.props.isAuthenticated);
    };

    render() {
        return (
            <div className='row'>
                <div className='col-sm'>
                    <nav className='navbar navbar-expand-lg navbar-light bg-light'>
                        <Link className='navbar-brand' to='/'>
                            Tasks App
                        </Link>
                        <button
                            className='navbar-toggler'
                            type='button'
                            data-toggle='collapse'
                            data-target='#navbarNav'
                            aria-controls='navbarNav'
                            aria-expanded='false'
                            aria-label='Toggle navigation'
                        >
                            <span className='navbar-toggler-icon'></span>
                        </button>
                        <div
                            className='collapse navbar-collapse'
                            id='navbarNav'
                        >
                            <ul id='nav-mobile' className='navbar-nav'>
                                <li className='nav-item'>
                                    <NavLink className='nav-link' to='/'>
                                        Home
                                    </NavLink>
                                </li>
                                <li className='nav-item'>
                                    <NavLink className='nav-link' to='/tasks'>
                                        Tasks
                                    </NavLink>
                                </li>
                                <li className='nav-item'>
                                    <NavLink className='nav-link' to='/profile'>
                                        Profile
                                    </NavLink>
                                </li>

                                {!this.props.isAuthenticated.auth && (
                                    <li className='nav-item'>
                                        <NavLink
                                            className='nav-link text-success'
                                            to='/login'
                                        >
                                            Login
                                        </NavLink>
                                    </li>
                                )}
                            </ul>
                            {this.props.isAuthenticated.auth && (
                                <button
                                    className='btn btn-sm btn-danger ml-auto'
                                    onClick={this.onLogoutclick}
                                >
                                    Logout
                                </button>
                            )}
                        </div>
                    </nav>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state: ReduxStoreState) => {
    return {
        isAuthenticated: state.isAuthenticated
    };
};

export default connect(mapStateToProps, { checkAuthentication })(Header);
