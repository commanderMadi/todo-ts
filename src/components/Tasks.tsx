import React from 'react';
import { ReduxStoreState } from '../reducers/';
import { connect } from 'react-redux';
import { Task, Category, getCategories, deleteTask, addTask } from '../actions';
import { AddTaskForm } from '../components/AddTaskForm';

export interface TasksProps {
    categories: Category[];
    getCategories: Function;
    deleteTask: typeof deleteTask;
    addTask: typeof addTask;
}

export class Tasks extends React.Component<TasksProps> {
    constructor(props: TasksProps) {
        super(props);

        this.onSubmit = this.onSubmit.bind(this);
    }
    componentDidMount() {
        this.props.getCategories();
    }

    onDelete(id: string) {
        this.props.deleteTask(id);
    }

    onSubmit(newTask: Task) {
        this.props.addTask(newTask);
    }

    renderCategories() {
        return (
            this.props.categories &&
            this.props.categories.map((category: Category, index: number) => {
                return (
                    <div className='col-md-6' key={index}>
                        <h3>{category.name}</h3>

                        {category.tasks.map((task: Task) => {
                            return (
                                <div className='mb-3' key={task.id}>
                                    <span className='pr-4'>{task.title}</span>
                                    <button
                                        className='task_remove btn btn-sm btn-danger'
                                        onClick={() => this.onDelete(task.id)}
                                    >
                                        Remove Task
                                    </button>
                                </div>
                            );
                        })}
                    </div>
                );
            })
        );
    }
    render() {
        return (
            <div>
                <AddTaskForm
                    onSubmit={this.onSubmit}
                    categories={this.props.categories}
                />
                <div className='row mt-5'>
                    <div className='col-12 mb-4'>
                        <h2>Tasks Categories</h2>
                    </div>
                    {this.props.categories.length > 0 ? (
                        this.renderCategories()
                    ) : (
                        <p className='no_data'>No data</p>
                    )}
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state: ReduxStoreState) => {
    return {
        categories: state.categories
    };
};

export default connect(mapStateToProps, { getCategories, addTask, deleteTask })(
    Tasks
);
