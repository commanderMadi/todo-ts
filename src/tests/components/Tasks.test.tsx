import React from 'react';
import { Tasks } from '../../components/Tasks';
import { shallow, ShallowWrapper } from 'enzyme';
import { Task } from '../../actions';
import { ActionTypes } from '../../actions/actiontypes';
import { initialData } from '../fixture/testdata';

let newTask: Task, wrapper: ShallowWrapper<Tasks>;
beforeEach(() => {
    newTask = {
        id: 'default-id-of-dummy-data',
        title: 'Run for half an hour',
        categoryTitle: 'Sports',
        error: null
    };
    wrapper = shallow(
        <Tasks
            categories={initialData}
            getCategories={() => ({
                type: ActionTypes.getCategories
            })}
            deleteTask={id => ({
                type: ActionTypes.deleteTask,
                payload: { id }
            })}
            addTask={() => {
                return {
                    type: ActionTypes.addTask,
                    payload: newTask
                };
            }}
        />
    );
});

test('Tasks component should load with all default data', () => {
    expect(wrapper).toMatchSnapshot();
});

test('A task should be deleted when delete function is triggered', () => {
    const onDeleteSpy = jest.spyOn(Tasks.prototype, 'onDelete');

    wrapper
        .find('button.task_remove')
        .at(1)
        .simulate('click');
    expect(onDeleteSpy).toHaveBeenLastCalledWith(initialData[1].tasks[0].id);
});

/* Still working on this test */

// test('AddTaskForm should correctly adds a task to the Tasks page on submit', () => {
//     const onSubmit = jest.spyOn(Tasks.prototype, 'onSubmit');
//     wrapper.find('AddTaskForm').prop('onSubmit')!(newTask as any);
//     const { title, id } = newTask;
//     expect(onSubmit).toHaveBeenLastCalledWith({ title, id });
// });
