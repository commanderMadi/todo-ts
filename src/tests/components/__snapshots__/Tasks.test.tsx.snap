// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tasks component should load with all default data 1`] = `
<div>
  <h1>
    Tasks Component
  </h1>
  <AddTaskForm
    categories={
      Array [
        Object {
          "name": "Sports",
          "tasks": Array [
            Object {
              "id": "042ffd34-d989-321c-ad06-f60826172424",
              "title": "Play Hockey",
            },
          ],
        },
        Object {
          "name": "Work",
          "tasks": Array [
            Object {
              "id": "109156be-c4fb-41ea-b1b4-efe1671c5836",
              "title": "Deliver Article",
            },
          ],
        },
      ]
    }
    onSubmit={[Function]}
  />
  <div
    key="0"
  >
    <h2>
      Sports
    </h2>
    <div
      key="042ffd34-d989-321c-ad06-f60826172424"
    >
      <h4>
        Play Hockey
      </h4>
      <button
        className="task_remove"
        onClick={[Function]}
      >
        Remove Task
      </button>
    </div>
  </div>
  <div
    key="1"
  >
    <h2>
      Work
    </h2>
    <div
      key="109156be-c4fb-41ea-b1b4-efe1671c5836"
    >
      <h4>
        Deliver Article
      </h4>
      <button
        className="task_remove"
        onClick={[Function]}
      >
        Remove Task
      </button>
    </div>
  </div>
</div>
`;
